### ISSUE TITLE
A way to persist precomputed grammars

### ISSUE THREAD

**Issue by dabrahams** [jeffreykegler/libmarpa#113]
When using MARPA to create a traditional compiler, I don't want to pay to build and precompute a grammar for each input file.  There should be a way to write out tables that allow a grammar to be quickly reconstructed with minimal data movement.  In particular, it would be ideal if these tables could be captured in statically-initialized data (e.g. global C arrays) and merely “pointed-at” by a new recognizer.

**Comment by jeffreykegler** [comment]
In the early days, the Marpa community looked carefully at this issue.  Precomputation is fast, and it seemed likely that the difference between precomputation and de-serialization of a serialized precomputed grammar would not be measurable.  Optimization is a moving target, and we won't really know the exact gain until we implement the serialization.  But it has an upper bound in the pre-computation time, less the time required to read a file whose size is proportional to the size of the precomputed grammar in its internal form.  This suggests that the potential gain is not likely to repay the time and effort required to make the experiment.

**Comment by dabrahams** [comment]
> On May 24, 2022, at 12:12 PM, Jeffrey Kegler ***@***.***> wrote:
> 
> In the early days, the Marpa community looked carefully at this issue. Precomputation is fast, and it seemed likely that the difference between precomputation and de-serialization of a serialized precomputed grammar would not be measurable.
> 
It’s not just precomputation but also all the calls to build rules and symbols.  That may still not account for much time in practice, of course; I don’t know, and I haven’t measured.  

That said, I’m not suggesting deserialization, which, given I/O and dynamic allocation costs, seems unlikely to produce a speedup.  Instead, I’m suggesting the entire precomputed grammar could be encoded in immutable data that’s part of the program image, in a format that can be used directly—by pointing at it.

**Comment by jeffreykegler** [comment]
Within the same run, you can achieve that effect by creating more than one recognizer with the same base grammar.  With this it's quite possible to run two (or more) parses side by side and the precomputation is done only once.

There are some limits because a couple of things are shared by all objects with the same base grammar.  One is the error code, and the other is the event queue.  The idea behind both is that both are supposed to be retrieved immediately after the call that generates them  This does imply that you cannot run recognizers with different base grammars in different threads.

This may still not be what you have in mind.  I know in olden days, there were UNIX programmers who would drop cores at arbitrary points and then re-purpose them as executables, reusing all the work done so far.  And I've toyed with the idea of leveraging pre-complied Lua chunks to similar effect.

**Comment by dabrahams** [comment]
> On May 24, 2022, at 2:56 PM, Jeffrey Kegler ***@***.***> wrote:
> 
> 
> Within the same run, you can achieve that effect by creating more than one recognizer with the same base grammar.
> 
If I was going to parse multiple files in a single run, I’d want to do it in parallel… which doesn’t work because the grammar is shared mutable state for all the recognizers operating on it.
> This may still not be what you have in mind. I know in olden days, there were UNIX programmers who would drop cores at arbitrary points and then re-purpose them as executables, reusing all the work done so far. And I've toyed with the idea of leveraging pre-complied Lua chunks to similar effect.
> 

The simplest approach, and what I had in mind, is to eliminate all internal pointers from the grammar data structure and instead use indices into arrays.

**Comment by jeffreykegler** [comment]
Not sure I follow, but there are no pointers from the grammar objects to their child objects.  The reason I don't consider sharing the grammar objects thread-safe is that mutable data in the base grammar object is shared among its child objects.  All of this mutable data could be eliminated without much trouble, with one exception: the reference counting for memory management.

I advertise C89 portability, and C89, if I am correct, neither guarantees the kinds of atomicity, nor offers the kind of locking that would allow me to claim that the reference counting is thread-safe.

**Comment by dabrahams** [comment]
> On May 25, 2022, at 8:06 AM, Jeffrey Kegler ***@***.***> wrote:
> 
> 
> Not sure I follow, but there are no pointers from the grammar objects to their child objects.
> 
I don’t know what MARPA’s internal data structures look like, but…

If there are any pointers between child objects or back to the grammar they would need to be eliminated too.  A grammar constructed as I’m suggesting would probably be a collection of arrays, with “child objects” stored in the array elements and any “child object pointers” replaced by array indices.  
> I advertise C89 portability, and C89, if I am correct, neither guarantees the kinds of atomicity, nor offers the kind of locking that would allow me to claim that the reference counting is thread-safe.
> 
I would handle that by adding three things:
1. A marpa_g_new_precomputed(a1, a2, …) constructor that builds a grammar around already-allocated arrays a1…aN, and also makes marpa_g_ref/marpa_g_unref into no-ops for the result.
2. A marpa_g_new_uncounted() that makes a grammar for which marpa_g_ref and marpa_g_unref are no-ops.    It becomes the users responsibility to call marpa_g_destroy(), or not, at the right time.
3. A marpa_g_destroy(g) function that unconditionally destroys/deallocates the arrays stored in the grammar

Sure, these operations are unsafe if used improperly, but that’s also true of marpa_g_unref.

**Comment by jeffreykegler** [comment]
I'll summarize the "how to" aspects of threads in this comment, and address "Is it worth it?" in a second comment.

With elimination of a few pieces of data shared with the child objects, we can create a Libmarpa which, after precomputation, has only one shared mutable -- the grammar's reference count.  We can eliminate this along the lines of a suggestion from @dabrahams:

In the configuration, have a "mutable" boolean.  `marpa_g_new()` copies this to a mutable flag.  If the mutable flag is unset, `marpa_g_ref()` and `marpa_g_unref()` are no-ops.  It then becomes necessary to "manually" destroy the grammar once threading ends, which may be done with a new `marpa_g_destroy()` method.

**Comment by jeffreykegler** [comment]
My opinions on how worthwhiile these changes to allow thread-safe base grammars would be are mainly informed from my C programming of threaded apps, and the disappointing results in the Perl world.  In C threads are great when you can get them working, but they're hard to debug.  A significant part of my income in my consulting days was rewriting threaded apps someone else had failed to get working.  All this cost the client, so the cost vs. risk vs. benefit requires a very important application.

In the Perl world, a lot of time and brainpower has been devoted to threading, with little to show for it AFAICT.  Marpa::R2 is thread-safe in the same sense as Libmarpa (no shared of base grammars) but I don't know of anyone who actually uses that.

I tried to find out how things are in the Swift world and found this:
https://www.hackingwithswift.com/quick-start/concurrency/understanding-threads-and-queues , which seems to counsel moderation.

My impression is that the current thread-safety of Libmarpa, which does not allow the sharing of base grammars, is adequate for the near or even forseeable future.

**Comment by dabrahams** [comment]
I don't know if you're trying to make the point that nobody should want to do stuff in parallel, but that's a pretty outdated view.  Swift now has extensive in-language support for concurrency, and IIRC the Swift compiler itself parses multiple files concurrently.  

Regardless, the point of these suggestions is to allow persistence and super-cheap construction.  They also happen to allow thread-safety.  Note that without this capability, MARPA will be good for my purposes as an experimental tool for a while, but (unless my project fizzles) I will certainly end up migrating to my own parser implementation.

**Comment by jeffreykegler** [comment]
If you create an experimental tool, then we will have a means to measure the cost of pre-computing the grammar for each thread.  Given an actual example of use, an optimization based on good measurements is likely to quickly move out of the wishlist and straight to the top of the priority queue.

**Comment by jddurand** [comment]
> This does imply that you cannot run recognizers with different base grammars in different threads.

I thought that the rule was: accessing a grammar G or a child, grand-child, etc. of it, must run synchroneously.
In other words, if I have different grammars G1 and G2, threads safety is ok as long as the threads do no execute concurrent calls on these grammars (or any child, grand-child etc of them -  this mean recognizers, bocage, valuation and so on) (?)

**Comment by jeffreykegler** [comment]
@jddurand The line you quote in https://github.com/jeffreykegler/libmarpa/issues/113#issuecomment-1199594793 (which is from https://github.com/jeffreykegler/libmarpa/issues/113#issuecomment-1136468145)  was a misstatement.  I've corrected it to

> This does imply that you cannot safely run two different recognizers with the same base grammar in different threads.

**Comment by jeffreykegler** [comment]
I'll attempt to summarize my understanding of where this discussion went.  I think the point turned out to be, not so much whether the particular thread-friendly change originally suggested should be made, but whether it could be the first step in making Libmarpa concurrency-safe in the modern sense.  Libmarpa IMHO is not "outdated" (https://github.com/jeffreykegler/libmarpa/issues/113#issuecomment-1140340762) in terms of current practice, but admittedly it falls well short of providing the kind of support for concurrency expected in new languages.

If I were to make adding concurrency safety my top short term priority, I would look long and hard at rewriting Libmarpa in a language which offers support for this, such as Swift or Rust.  There are trade-offs, but leaving behind a language whose concurrency and memory management support is as impoverished as is the case with C language, is certainly tempting.  @dabrahams has decided to do undertake a Swift clone of the Marpa/Earley/Leo algorithm, and I look with interest to the results.

My top short term priority is supportng the current generation of users and applications.  After that I have some ideas for adding new capabilities to the Marpa algorithm.  It is not impossible these will wind in pull requests submitted to a Swift clone of Marpa.

**Comment by jeffreykegler** [comment]
@dabrahams: I have added a comment (https://github.com/jeffreykegler/libmarpa/issues/113#issuecomment-1202993256) that is intended to summarize the result of this discussion from my POV.

Since, for the reasons given in that comment, I plan no near term action on this request, I would like to close this issue.  Absent feedback to the contrary, I will do so after a couple of days.

**Comment by jeffreykegler** [comment]
Closed per https://github.com/jeffreykegler/libmarpa/issues/113#issuecomment-1202997380.

**Comment by dabrahams** [comment]
It's interesting that it was fast enough with the R2 front end and ESLIF is [embarrassingly faster](https://colabti.org/irclogger/irclogger_log/marpa?date=2022-08-05).  I wonder if your view of the relative cost of preprocessing would change if you adopted a faster front-end.

**Comment by jeffreykegler** [comment]
Numbers could change my POV, yes.

**Comment by jeffreykegler** [comment]
Could you point me to something that works this way?

I'm particularly confused about how an array vs. pointer distinction might confer thread-safety in the C context. A C array, if referred to via a bareword, is a pointer. And indexing a C array is literally synonymous with pointer arithmetic. And all this is before optimization.

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Comment by dabrahams** [comment]
> On Oct 11, 2022, at 12:13 AM, Jeffrey Kegler ***@***.***> wrote:
> 
> I'm particularly confused about how an array vs. pointer distinction might confer thread-safety in the C context. A C array, if referred to via a bareword, is a pointer. And indexing a C array is literally synonymous with pointer arithmetic. And all this is before optimization.

That wasn’t being suggested in order to achieve thread safety, but to achieve address independence of the representation, so it could be created by one parser-generation executable and persisted as arrays in a generated source file that is compiled into the executable doing the actual parsing.  That’s how Yacc and Bison work, for example.  

This array transformation can in fact be leveraged to achieve memory safety, though, because you can encapsulate every array access in a safe function that does bounds checking.  Achieving complete thread-safety of a C API that does mutation is probably nearly impossible, but you can do a LOT to support practical multithreaded usage without the massive copying/regeneration that’s currently required.  Probably the simplest way is to have the grammar post-processing step effectively empty all copies of the incoming mutable grammar type used for grammar construction and return a new immutable grammar type that has taken ownership of the incoming grammar’s contents and that can be used for parsing.  That would amount to manually encoding the semantics of a linear type system into your API.

**Comment by jeffreykegler** [comment]
@dabrahams Thanks!

I've done a bit of reading since this thread began.  I wonder if, rather than try to make C do something it is not really designed to do, it is not better to rewrite in a thread-safety-friendly language.  If I understand correctly, Rust and  Swift are two of these.

**Comment by dabrahams** [comment]
First, let me repeat:

> the point of these suggestions is to allow persistence and super-cheap construction. They also happen to allow thread-safety.

Second, regarding a change of language: 🤷 
1. In Rust you'd end up doing something similar to what I suggested above for C.
2. In Swift you *could* easily keep everything mutable all the time and have thread safety, but it would still be a much better design to do in Swift what I suggested above for C.
3. Both of these languages create interoperability barriers with other popular languages, so be aware of what you're losing by leaving C.
4. If you *were* going to make the change, I'd suggest Swift; you'll be less prone to get tangled up in lifetime annotations and borrow checking.
